build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -0,0 +1,154 @@
+ * limitations under the License.
+ */
+
+const BAIL_OUT_CONDITIONS = {
+  // If this isn't a FunctionDeclaration, bail out on modification.
+  isNotFunction: (t, path) => !t.isFunctionDeclaration(path.node),
+
+  // If this FunctionDeclaration is a Generator, bail out on modification.
+  isGenerator: (t, path) => path.node.generator,
+
+  // If this FunctionDeclaration doesn't have a single return statement, bail out on modification
+  isNotSingleReturnStatment: (t, path) => {
+    if (!path.get('body').isBlockStatement()) {
+      return false;
+    }
+    const body = path.get('body.body.0');
+    return !body || !body.isReturnStatement();
+  },
+
+  // Since we don't know if exported members are 'new'd, bail out on modification.
+  isExported: (t, path) =>
+    path.findParent(
+      path =>
+        path.isExportNamedDeclaration() || path.isExportDefaultDeclaration()
+    ),
+
+  // If the function contains usage of `arguments`, bail out on modification.
+  containsArgumentsUsage: (t, path) => {
+    let containsArgumentsUsage = false;
+    path.traverse({
+      Identifier(path) {
+        if (t.isIdentifier(path.node, {name: 'arguments'})) {
+          containsArgumentsUsage = true;
+          path.stop();
+        }
+      },
+    });
+    return containsArgumentsUsage;
+  },
+
+  // If the function contains a ThisExpression, converting from a FunctionDeclaration to a
+  // VariableDeclaration => VariableDeclarator => ArrowFunctionExpression isn't necessarily valid.
+  // Bail out on modification.
+  containsThisExpression: (t, path) => {
+    let containsThis = false;
+    path.traverse({
+      ThisExpression() {
+        containsThis = true;
+        path.stop();
+      },
+    });
+    return containsThis;
+  },
+
+  // If the FunctionDeclaration identifier is newed in the scope of this program, bail out on modification.
+  isNewedInProgramScope: (t, path) => {
+    const {name} = path.get('id').node;
+    let isNewed = false;
+    path
+      .findParent(path => path.isProgram())
+      .traverse({
+        NewExpression(path) {
+          if (t.isIdentifier(path.node.callee, {name})) {
+            isNewed = true;
+            path.stop();
+          }
+        },
+      });
+    return isNewed;
+  },
+};
+
+function createVariableDeclaration(t, path) {
+  const {params, body, async, id} = t.cloneNode(path.node);
+  const arrowFunction = t.arrowFunctionExpression(
+    params,
+    body.body[0].argument,
+    async
+  );
+  const declarations = t.variableDeclarator(
+    t.identifier(id.name),
+    arrowFunction
+  );
+
+  // Making an existing FunctionDeclaration a let VariableDeclaration is safe.
+  // The name was already reserved for the FunctionDeclaration.
+  return t.variableDeclaration('let', [declarations]);
+}
+
+// Attempt to convert simple single ReturnStatement FunctionDeclarations to ArrowFunctionExpressions.
+// See BAIL_OUT_CONDITIONS for reasons why FunctionDeclarations would not be modified.
```

Comment:
would you mind adding a negative test that it doesn't transform `FunctionExpression`'s , worried it'll conflict with the non newable factory functions again. i know it wouldn't since the only visitor we have is `FunctionDeclaration` but maybe somewhere down the line somebody thinks its a good idea to transform those too and doesn't know why we wouldn't want to

Reply:
Good idea, I'll add this in.

Reply:
example below. These passed in `FunctionExpression` to these `register*` functions are eventually called `new` on.

```
registerServiceBuilder(win, 'story-analytics', function() {
  return new StoryAnalyticsService();
});
```

Reply:
Let's make this a bail out condition, if a function is an argument to a method, it cannot be transformed since we dont know the accepting function will `new` or not the passed argument function.

Reply:
Addressed with a test in ecab0f5.